### DISCUSSION TITLE
change between last lookup and final output

### DISCUSSION THREAD

**Discussion by bdenckla** [harfbuzz/harfbuzz#2962] [Q&A] [answered]
I have a string and font for which `hb-shape --trace` shows changes between the last (GPOS) lookup and the final output. Taking some liberties with whitespace to line things up, the last two lines of output are as follows:

```
trace: end lookup 53    buffer: [lamed=0+901|qamats=0@648,0+0|meteg=0@550,0+0|hiriq=0@-238,0+0|finalmem=5+1004]
                                [finalmem=5+1004|hiriq=0@312,0+0|meteg=0@550,0+0|qamats=0@648,0+0|lamed=0+901]
```
I see that the final change to RTL is part of these last set of changes, but even when I "undo" that by taking some liberties by reversing the final output, I get the following, showing a position change in `hiriq`, from -238 to 312:
```
trace: end lookup 53    buffer: [lamed=0+901|qamats=0@648,0+0|meteg=0@550,0+0|hiriq=0@-238,0+0|finalmem=5+1004]
                                [lamed=0+901|qamats=0@648,0+0|meteg=0@550,0+0|hiriq=0@312,0+0|finalmem=5+1004]
```
I was hoping to try to understand what lookup this final position of 312 comes from, but I'm a bit stuck, since this final position of 312 seems to come from something after the lookups, or at least after the lookups that are traced.

**Comment by behdad** [comment, marked as answer]
The hiriq is attached to the meteg. During lookup processing you don't see that in the trace output. At the end, the offsets are adjusted for attachments. 312 = 550 + -238.

**Reply by bdenckla** [reply]
Thanks @behdad that makes sense. Does `hb-shape` have an option to do "deeper" tracing to show these (mark-to-mark) adjustments? If not, do you think there is any interest (beyond my interest) in adding a feature that would allow `hb-shape` to do such "deeper" tracing?

**Reply by behdad** [reply]
That's desirable but we have not got to do that yet. See:
https://github.com/harfbuzz/harfbuzz/issues/358
https://github.com/harfbuzz/harfbuzz/issues/2468
